feat: Add support For SeedVR2 (CORE-6)

yousef-rafat opened the pull request:
<img width="1808" height="743" alt="Screenshot 2025-12-22 194424" src="https://github.com/user-attachments/assets/f521fafb-496d-4a71-9b92-df71ea13f4e3" />

yousef-rafat committed:
init

yousef-rafat committed:
add nodes

yousef-rafat committed:
continue building nodes / testing vae

yousef-rafat committed:
testing the model

yousef-rafat committed:
mostly fixing mistakes

yousef-rafat committed:
fixed some issues

yousef-rafat committed:
.

yousef-rafat committed:
testing

yousef-rafat committed:
..

yousef-rafat committed:
lora, 7b model, cfg

yousef-rafat committed:
vae fix

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
works

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

Kosinkadink commented:
Currently, trying either image or video inputs results in this error during VAE Decode:

```
!!! Exception during processing !!! Given groups=1, weight of size [512, 16, 3, 3, 3], expected input[1, 1, 16, 3, 332] to have 16 channels, but got 1 channels instead
Traceback (most recent call last):
  File "/home/kosin/ComfyUI/execution.py", line 516, in execute
    output_data, output_ui, has_subgraph, has_pending_tasks = await get_output_data(prompt_id, unique_id, obj, input_data_all, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                                                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/execution.py", line 330, in get_output_data
    return_values = await _async_map_node_over_list(prompt_id, unique_id, obj, input_data_all, obj.FUNCTION, allow_interrupt=True, execution_block_cb=execution_block_cb, pre_execute_cb=pre_execute_cb, v3_data=v3_data)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/execution.py", line 304, in _async_map_node_over_list
    await process_inputs(input_dict, i)
  File "/home/kosin/ComfyUI/execution.py", line 292, in process_inputs
    result = f(**inputs)
             ^^^^^^^^^^^
  File "/home/kosin/ComfyUI/nodes.py", line 298, in decode
    images = vae.decode(samples["samples"])
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/custom_nodes/benchmark/__init__.py", line 307, in wrapper_VAE_decode
    return func(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/sd.py", line 799, in decode
    out = self.process_output(self.first_stage_model.decode(samples, **vae_options).to(self.output_device).float())
                              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 1564, in decode
    x = super().decode(latent).squeeze(2)
        ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 1455, in decode
    decoded = self.slicing_decode(z)
              ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 1486, in slicing_decode
    return self._decode(z)
           ^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 1479, in _decode
    output = self.decoder(latent)
             ^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 1290, in forward
    sample = self.conv_in(sample)
             ^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1773, in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/module.py", line 1784, in _call_impl
    return forward_call(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/comfy/ldm/seedvr/vae.py", line 386, in forward
    return super().forward(input)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/conv.py", line 717, in forward
    return self._conv_forward(input, self.weight, self.bias)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/kosin/ComfyUI/venv/lib/python3.12/site-packages/torch/nn/modules/conv.py", line 712, in _conv_forward
    return F.conv3d(
           ^^^^^^^^^
RuntimeError: Given groups=1, weight of size [512, 16, 3, 3, 3], expected input[1, 1, 16, 3, 332] to have 16 channels, but got 1 channels instead

```

yousef-rafat committed:
fix vae issue

yousef-rafat committed:
.

yousef-rafat committed:
added var length attention and fixed the vae issue

yousef-rafat committed:
...

yousef-rafat committed:
improvements

yousef-rafat committed:
video works

yousef-rafat committed:
remove cfg cutoff node

yousef-rafat committed:
.

yousef-rafat committed:
revert file perm.

yousef-rafat committed:
..

yousef-rafat committed:
ruff

yousef-rafat committed:
removed print statement

yousef-rafat committed:
fixed the speed issue

yousef-rafat committed:
outputs/speed/memory match custom node

yousef-rafat committed:
final changes

yousef-rafat committed:
.

yousef-rafat committed:
added other types of attention + compatibility
with images

yousef-rafat committed:
tile edge case handles by padding vid

yousef-rafat committed:
ruff

Kosinkadink committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
fixed manual vae loading

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
rope, attetntion update | vae on cpu warning

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
ruff

yousef-rafat committed:
.

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
Merge branch 'master' into seedvr2

Kosinkadink committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
fp16 check + changed to lab

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
neg condition was passed as context

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
custom rmsnorm

yousef-rafat committed:
.

yousef-rafat committed:
7b specific

yousef-rafat committed:
simplified var attn

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
.

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
tested with a modern version of comfyui

yousef-rafat committed:
...

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
output reshape

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat committed:
.

Kosinkadink commented:
Some thoughts:
1. var length attention implementation seems fairly. Would need @comfyanonymous to give his take.
2. I see there is some rope re-implementation and some rope reuse as well. Is some of the rope stuff for SeedVR2 unique enough that it cannot be refactored in a way that reuses apply_rope1?
3. ldm/seedvr/vae.py should not import from tiled_vae from comfy_extras.nodes_seedvr. If comfy_extras import fails for some reason, all of ComfyUI would crash. The function should be moved elsewhere and then nodes_seedvr can import it.
4. There are some odd casts of types + looking up comfy's torch device while in forward functions: https://github.com/Comfy-Org/ComfyUI/blob/553f71aa9e39cab8ac00137a2172f0cf076bccb8/comfy/ldm/seedvr/model.py#L624-L629 https://github.com/Comfy-Org/ComfyUI/blob/553f71aa9e39cab8ac00137a2172f0cf076bccb8/comfy/ldm/seedvr/model.py#L826-L829 https://github.com/Comfy-Org/ComfyUI/blob/553f71aa9e39cab8ac00137a2172f0cf076bccb8/comfy/ldm/seedvr/model.py#L869-L870 https://github.com/Comfy-Org/ComfyUI/blob/553f71aa9e39cab8ac00137a2172f0cf076bccb8/comfy/ldm/seedvr/model.py#L1195-L1200 https://github.com/Comfy-Org/ComfyUI/blob/553f71aa9e39cab8ac00137a2172f0cf076bccb8/comfy/ldm/seedvr/model.py#L1415-L1419

It seems a bit un-comfy like. Did you encounter some device mismatch issues? If so, those should be dealt with in such a way that the passed in ops, load_device, etc. things into the model resolve with with the model management system.

yousef-rafat committed:
Merge branch 'master' into seedvr2

yousef-rafat committed:
pr fixes

yousef-rafat committed:
Merge branch 'seedvr2' of https://github.com/yousef-rafat/ComfyUI into seedvr2

yousef-rafat commented:
Fixed 3, 4.
For 2, it already uses apply_rope1. I just refactored some parts of the rope implementation as a matrix to work with apply_rope1
You can see it in `apply_rotary_emb`
